### ISSUE TITLE
Find best hyperparameters with pipeline /  02_end_to_end_machine_learning_project.ipynb

### ISSUE THREAD

**Issue by quidba7** [ageron/handson-ml#201]
Hi Aurelien,
Thank you for this amazing book.

I'm working on simpler code to chose the best parameters for feature selection / filling NaN / manager outliers / Standardization. 

I need to understand how the GridSearch in Line 128, Question 4 recognizes that 'preparation__num_pipeline__imputer__strategy'  from param _grid is actually a parameter of the num_pipeline Imputer method defined earlier line 73 ? ( of which full_pipeline is derived ) 
> ['mean','median', 'most_frequent'] 

 There is no parameter with such name, It was a step in the pipeline called 
>'imputer'

 am I missing a line of code ?

Many thks

**Comment by ageron** [comment]
Hi @quidba7 ,
Thanks for your kind words! :)

Consider the following simple pipeline:

```python
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import Imputer, StandardScaler
pipeline = Pipeline([
    ("foo", Imputer()),
    ("bla", StandardScaler()),
    ("lin", LinearRegression()),
])
```

You can access the `Imputer`'s `strategy` hyperparameter in several ways.

First, you can get a handle on the `Imputer` instance via the pipeline's `steps` instance variable:

```pycon
>>> pipeline.steps[0]
('foo', Imputer(axis=0, copy=True, missing_values='NaN', strategy='mean', verbose=0))
>>> pipeline.steps[0][1].strategy
'mean'
```

Or you can use the `named_steps` instance variable, which just returns a convenient `dict` of the `steps`:

```pycon
>>> pipeline.named_steps["foo"].strategy
'mean'
```

Finally, you can use `get_params()`, which returns the full list of hyperparameters of all the estimators in the pipeline, prefixed with their name plus two underscores:

```pycon
>>> pipeline.get_params()["foo__strategy"]
'mean'
>>> pipeline.get_params()
{'memory': None, 'steps': [('foo', Imputer(axis=0, copy=True, missing_values='NaN', strategy='mean', verbose=0)), ('bla', StandardScaler(copy=True, with_mean=True, with_std=True)), ('lin', LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False))], 'foo': Imputer(axis=0, copy=True, missing_values='NaN', strategy='mean', verbose=0), 'bla': StandardScaler(copy=True, with_mean=True, with_std=True), 'lin': LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False), 'foo__axis': 0, 'foo__copy': True, 'foo__missing_values': 'NaN', 'foo__strategy': 'mean', 'foo__verbose': 0, 'bla__copy': True, 'bla__with_mean': True, 'bla__with_std': True, 'lin__copy_X': True, 'lin__fit_intercept': True, 'lin__n_jobs': 1, 'lin__normalize': False}
```

Now let's nest this pipeline within another pipeline:

```python
master_pipeline = Pipeline([
    ("baz", StandardScaler()),
    ("my_sub_pipe", pipeline)
])
```

The same three approaches are possible with this new pipeline:

>>> master_pipeline.steps[1][1].steps[0][1]
Imputer(axis=0, copy=True, missing_values='NaN', strategy='mean', verbose=0)
>>> master_pipeline.steps[1][1].steps[0][1].strategy
'mean'
>>> master_pipeline.named_steps["my_sub_pipe"].named_steps["foo"].strategy
'mean'
>>> master_pipeline.get_params("my_sub_pipe__foo__strategy")
{'memory': None, 'steps': [('baz', StandardScaler(copy=True, with_mean=True, with_std=True)), ('my_sub_pipe', Pipeline(memory=None,
     steps=[('foo', Imputer(axis=0, copy=True, missing_values='NaN', strategy='mean', verbose=0)), ('bla', StandardScaler(copy=True, with_mean=True, with_std=True)), ('lin', LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False))]))], 'baz': StandardScaler(copy=True, with_mean=True, with_std=True), 'my_sub_pipe': Pipeline(memory=None,
     steps=[('foo', Imputer(axis=0, copy=True, missing_values='NaN', strategy='mean', verbose=0)), ('bla', StandardScaler(copy=True, with_mean=True, with_std=True)), ('lin', LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False))]), 'baz__copy': True, 'baz__with_mean': True, 'baz__with_std': True, 'my_sub_pipe__memory': None, 'my_sub_pipe__steps': [('foo', Imputer(axis=0, copy=True, missing_values='NaN', strategy='mean', verbose=0)), ('bla', StandardScaler(copy=True, with_mean=True, with_std=True)), ('lin', LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False))], 'my_sub_pipe__foo': Imputer(axis=0, copy=True, missing_values='NaN', strategy='mean', verbose=0), 'my_sub_pipe__bla': StandardScaler(copy=True, with_mean=True, with_std=True), 'my_sub_pipe__lin': LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False), 'my_sub_pipe__foo__axis': 0, 'my_sub_pipe__foo__copy': True, 'my_sub_pipe__foo__missing_values': 'NaN', 'my_sub_pipe__foo__strategy': 'mean', 'my_sub_pipe__foo__verbose': 0, 'my_sub_pipe__bla__copy': True, 'my_sub_pipe__bla__with_mean': True, 'my_sub_pipe__bla__with_std': True, 'my_sub_pipe__lin__copy_X': True, 'my_sub_pipe__lin__fit_intercept': True, 'my_sub_pipe__lin__n_jobs': 1, 'my_sub_pipe__lin__normalize': False}
>>> master_pipeline.get_params()["my_sub_pipe__foo__strategy"]
'mean'
```

**Comment by quidba7** [comment]
Thank you so much for taking the time to write a great explanation.
Now back on track for building intelligent systems !
